lib/xenops_server.ml

```diff
@@ -804,7 +804,14 @@ let simplify f =
 
 let rec atomics_of_operation = function
 	| VM_start id ->
-		[
+		(
+			let module B = (val get_backend () : S) in
+			let power = (B.VM.get_state (VM_DB.read_exn id)).Vm.power_state in
+			begin match power with
+				| Running -> raise (Bad_power_state(Running, Halted))
```

Comment:
I agree that `atomics_of_operation` is not the right place, because purpose of the function is to break down a high-level operation into a list of sub-operations.

Furthermore, I would prefer to make operations such as VM_start idempotent. For VM_start, that would mean that it should just return without error if the VM is already running. Ideally, we would make the sub-operations of VM_start (VM_create, VM_build, VBD.plug, ...) idempotent in the backend itself (some of them may be idempotent already).